partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprSubquery.kt

```diff
@@ -0,0 +1,83 @@
+
+import org.partiql.errors.TypeCheckException
+import org.partiql.eval.internal.Record
+import org.partiql.eval.internal.operator.Operator
+import org.partiql.value.PartiQLValue
+import org.partiql.value.PartiQLValueExperimental
+import org.partiql.value.StructValue
+import org.partiql.value.check
+import org.partiql.value.listValue
+import org.partiql.value.nullValue
+import java.util.Stack
+
+/**
+ * The PartiQL Specification talks about subqueries and how they are coerced. Specifically, subqueries are
+ * modeled as a COLL_TO_SCALAR(SELECT VALUE <tuple> ...) where the SELECT VALUE must return a single row containing
+ * a TUPLE.
+ *
+ * @see [getValues]
+ */
+internal abstract class ExprSubquery : Operator.Expr {
+
+    abstract val constructor: Operator.Expr
+    abstract val input: Operator.Relation
+    abstract val scopes: Stack<Record>
+
+    internal class Row(
+        override val constructor: Operator.Expr,
+        override val input: Operator.Relation,
+        override val scopes: Stack<Record>
+    ) : ExprSubquery() {
+        @PartiQLValueExperimental
+        override fun eval(record: Record): PartiQLValue {
+            val values = getValues(record) ?: return nullValue()
+            return listValue(values.asSequence().toList())
+        }
+    }
+
+    internal class Scalar(
+        override val constructor: Operator.Expr,
+        override val input: Operator.Relation,
+        override val scopes: Stack<Record>
+    ) : ExprSubquery() {
+        @PartiQLValueExperimental
+        override fun eval(record: Record): PartiQLValue {
+            val values = getValues(record) ?: return nullValue()
+            if (values.hasNext().not()) {
+                throw TypeCheckException()
+            }
+            val singleValue = values.next()
+            if (values.hasNext()) {
+                throw TypeCheckException()
+            }
+            return singleValue
+        }
+    }
+
+    /**
+     * This grabs the first row of the [input], asserts that the [constructor] evaluates to a TUPLE, and returns an
+     * [Iterator] of the [constructor]'s values. These values are then used by [Scalar] and [Row].
+     *
+     * @return an [Iterator] of the [constructor]'s values of the first row from the [input]. Returns null when
+     * no rows are returned from the [input].
+     * @throws TypeCheckException when more than one row is returned from the [input].
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    fun getValues(record: Record): Iterator<PartiQLValue>? {
+        scopes.push(record)
+        input.open()
+        val firstRecord = input.next()
+        if (firstRecord == null) {
+            scopes.pop()
+            return null
+        }
+        val tuple = constructor.eval(firstRecord).check<StructValue<*>>()
+        val secondRecord = input.next()
+        scopes.pop()
+        if (secondRecord != null) {
+            throw TypeCheckException()
+        }
+        return tuple.values.iterator()
```

Comment:
Why do we want a tuple.values.iterator()?

Why not just

```
private fun Operator.Relation.first(): StructValue<*>?

// let n = # rows
//   n = 0      ->  return null
//   n = 1       ->  return constructor.eval(row[0])
//   n > 1       -> throw CardinalityException()   
```

I don't think we want a TypeCheckException because that might be considered MISSING. The specification suggest we throw a cardinality exception for n > 1 in strict and ignore in permissive.

> An PartiQL extension with respect to SQL is that, in the permissive mode, subqueries that fail to coerce to the
required type (scalar or tuple) still run, as opposed to failing. They simply omit from the results the data that
correspond to the coercion failures.

Also, we'll need to throw/ignore for scalar/row coercions depending on the number of struct values.

Reply:
Further on in the same section:
> In this example, in one instance the inner SELECT evaluates to a collection with more than one element.
Because the COLL_TO_SCALAR function produces a MISSING instead of failing, the query works

Either way, in permissive the error should be coerced to MISSING. I can make it a cardinality exception like the SQL:1999 spec says, but I'll just need to coerce that too.

Reply:
And as for why the iterator: it doesn't really matter. In both the Scalar and Row scenarios, you need access to the values, so it was just for de-duplicating one or two lines of code.